src/cdn_definitions/data.yaml

```diff
@@ -251,3 +251,175 @@ repo_overrides:
       if_created_after: "2023-06-01T14:00:00Z"
       key: example
       value: true
+
+# directory_layouts is used to define DirectoryLayout objects and consist of a
+# pattern, to match against content URL, and attributes for the resulting
+# object, often present as regex groups in the pattern.
+#
+# Allowed value types; string, bool, null
+#
+# With the exception of "rhelalt", which is expected to be stored as bool,
+# a "true" value is used to signal that regex matching should be applied for the
+# corresponding group in the pattern.
+#
+# E.g.;
+#   `type: true`    ==  `type = re.match(...).group("type")`
+#   `type: false`   ==  `type = False`
+#   `type: "dist"`  ==  `type = "dist"`
+directory_layouts:
+  - name: DirectoryLayoutC
+    # "/content/<type>/<platform>/<rhui>/<variant>/<major>.<minor>/<arch>/<rest>"
+    pattern: |-
+      (?x)^
+      /
+      (?:content|shadow)/
+      (?P<type>aus|dist|beta|(?:retired-)?els|eus|htb|rhb|tus|e4s)/
+      (?P<platform>rhes|rhs|cf-me)/
+      (?P<rhui>rhui/)?
+      (?P<variant>[^/]+)/
+      (?P<major_version>[0-9])\.(?P<minor_version>[0-9]+)/
+      (?P<arch>[^/]+)/
+      (?P<rest>.+)$
```

Comment:
Is there a plan right now how to deal with `parse_rest`?
I'm not sure the schema here is going deep enough right now.

What I mean is that there is currently a `ProductContentSet.parse_rest` which further decomposes the `rest` portion and derives some attributes from it.  This parsing has also been overridden once, in the SUSE layout.

Ideally, all the logic currently happening in `parse_rest` is somehow expressed in the layout definitions instead.
If this is not done then it's not really practical to use the definitions outside of cdn-utils (or at least not without duplicating the code).
Also, even if the definitions are written such that they only work with cdn-util's `parse_rest`, code using the definitions needs to somehow decide whether to use the base implementation or SUSE-specific.

Reply:
Well, these definitions can only work with cdn-utils either way, as that's where they'll be turned into DirectoryLayout objects. And we can easily determine the suse behavior from the pattern or definition; currently SLESLayout specifies `platform: suse` and includes 'suse/' in the pattern besides.

It seems `parse_rest` is really just seeking to fill 'repo', 'layered_product', 'layered_product_version', and in the suse case, 'arch'. I'm no regex expert but it seems like it _should_ be possible to translate these to the pattern if we really must.

Reply:
I see. Yeah, that seems like a good idea to me. Done!